engine/consensus/matching/engine.go

```diff
@@ -900,7 +935,158 @@ func (e *Engine) requestPending() error {
 		Int("finalized_blocks_without_result", len(missingBlocksOrderedByHeight)).
 		Msg("requesting receipts")
 	for _, blockID := range missingBlocksOrderedByHeight {
-		e.requester.EntityByID(blockID, filter.Any)
+		e.receiptRequester.EntityByID(blockID, filter.Any)
+	}
+
+	return nil
+}
+
+// requestPendingApprovals requests approvals for chunks that haven't collected
+// enough approvals. When the number of unsealed finalized blocks exceeds the
+// threshold, we go through the entire mempool of incorporated-results, which
+// haven't yet been sealed, and check which chunks need more approvals. We only
+// request approvals if the block incorporating the result is below the
+// threshold.
+//
+//                                   threshold
+//                              |                   |
+// ... <-- A <-- A+1 <- ... <-- D <-- D+1 <- ... -- F
+//       sealed       maxHeightForRequesting      final
+func (e *Engine) requestPendingApprovals() error {
+
+	// Skip requesting approvals if they are not required for sealing.
+	// TODO: this is only here temporarily to ease the migration to new chunk
+	// based sealing.
+	if !e.requireApprovals {
+		return nil
+	}
+
+	// last sealed block
+	sealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get sealed height: %w", err)
+	}
+
+	// last finalized block
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized height: %w", err)
+	}
+
+	// only request if number of unsealed finalized blocks exceeds the threshold
+	log := e.log.With().
+		Uint64("finalized_height", final.Height).
+		Uint64("sealed_height", sealed.Height).
+		Uint64("approval_requests_threshold", e.approvalRequestsThreshold).
+		Logger()
+
+	maxHeightForRequesting := final.Height - e.approvalRequestsThreshold
```

Comment:
on startup `final.Height` would be `0`, would the calculation underflow? 

Maybe we could extract the calculation into a function, and add unit tests to the function:
```
func shouldRequest(sealedHeight, finalHeight, approvalRequestsThreshold uint64) bool 
```

Reply:
Thanks Leo. Good catch. 

Maybe we can just switch around the order of https://github.com/onflow/flow-go/blob/fb2b796cdb34719511833727db84627f75de75ac/engine/consensus/matching/engine.go#L983-L987 and use the check Leo suggest _first_:
```golang
	if sealed.Height + e.approvalRequestsThreshold >= final.Height {
		log.Debug().Msg("skip requesting approvals as number of unsealed finalized blocks is below threshold")
		return nil
	}
	// Reaching the following code implies: sealed.Height < final.Height - approvalRequestsThreshold
	// Hence, the following operation cannot underflow
	maxHeightForRequesting := final.Height - e.approvalRequestsThreshold


```